### DISCUSSION TITLE
What order is the search tree explored in?

### DISCUSSION THREAD

**Discussion by andrew222651** [pschanely/CrossHair#102] [Q&A] [answered]
In order to take a new execution path, CrossHair needs to visit a new node in the search tree. In what order are nodes visited, e.g. breadth first, depth first? And where in the code is this done?

**Comment by pschanely** [comment, marked as answer]
Great question. Path decisions are made randomly, guided by some heuristics. (though we avoid repeating the exact same paths)

Most commonly, the decision is [made here](https://github.com/pschanely/CrossHair/blob/9a7185ef2365a4319bfb046a5450f7c4761751c8/crosshair/statespace.py#L353). We bias towards picking False for booleans, as it more likely to break us out of loops, hopefully exploring shorter paths first.

The search tree will also include nodes that don't obviously correspond to branches in the code under analysis. If this is of interest, I can prioritize doing some sort of write-up about the search tree and the different kinds of nodes - let me know!

**Reply by andrew222651** [reply]
Thanks. More write-ups are always good!
